homeassistant/components/roborock/models.py

```diff
@@ -23,3 +28,28 @@ def as_dict(self) -> dict[str, dict[str, Any]]:
             "product": self.product.as_dict(),
             "props": self.props.as_dict(),
         }
+
+
+@dataclass
+class RoborockImageExtraStoredData(ExtraStoredData):
+    """A model to describe the maps for restore."""
+
+    map_flag: int
+    cached_image: bytes
+
+    def as_dict(self) -> dict:
+        """Return the image data as a dictionary."""
+        return {
+            "map_flag": self.map_flag,
+            "cached_image": base64.b64encode(self.cached_image).decode("utf-8"),
+        }
+
+    @staticmethod
+    def from_dict(
+        serialized_dict: dict,
+    ) -> RoborockImageExtraStoredData:
+        """Take the serialized dict and convert to object."""
+        return RoborockImageExtraStoredData(
+            map_flag=serialized_dict["map_flag"],
+            cached_image=base64.b64decode(serialized_dict["cached_image"]),
```

Comment:
Hm, this will make the restored entity cache huge as it now contains a multiple whole images in base64 (which takes up more space than original image).

Reply:
So from what it seems for me (on my map at least)

The response from the api has a length of 269612 bytes.

The cached map aka the image has a length of 7909 bytes

And the base64 encoded serialized image has a length of 10548 bytes.

I'm open to a better way of storing the image, I just wanted to make sure that I was able to consistently serialize and serialize the image.

Reply:
I think it's better stored on disk outside of the restore entity cache. That cache is written to disk every 15min. In my production the restore state cache is 123KB. That would increase with 25% (I have 3 maps). 

If you write it to disk, make sure it's cleaned up when the config entry is removed. Maybe you can make a folder inside `.storage`: `<config>/.storage/roborock_<config entry ID>/maps` ?

Reply:
That would be fine by me. My only question is how often do I write it to the disk? Should it only be written to the disk on config entry unloading? Or should it get done at the same point as the restore entity?

Reply:
I also remembered that the response from the api before we decompress it is a gzip file. So I could also save that to disk and that would save some space. I'm not 100% sure a clean way to do that as it is kind of in the middle of the process within our package, but if we want it to be even smaller, I can do that.

Reply:
I think firing off an async write when the map has changed is fine, and could include some rate limiting or debouncing (e.g. you don't necessarily need to write every time the robot moves).

I don't think its worth it to do anything fancy there in the middle of the library.  I've gone down that road when writing lots of objects where nest passes in a storage interface into the client library for saving images/videos [nest media source](https://github.com/home-assistant/core/blob/734a614eb8dbd2a7026c16ea1eaa2d52735bb510/homeassistant/components/nest/media_source.py#L97) and the [interface](https://github.com/allenporter/python-google-nest-sdm/blob/85216804caa5e993f00d68cf98b744fcddb5f48b/google_nest_sdm/event_media.py#L123)  but i don't think its worth it here.

Reply:
Need to do tests - but I took what you did in Nest and modified it. No injection within the library, but I did think it would be useful to have a central object for handling the storage.

I'll try to get tests done soon, but for now I'd appreciate any feedback on the implementation. Will mark as draft until I get tests done.